src/Compilers/Core/Portable/Operations/OperationMapBuilder.cs

```diff
@@ -0,0 +1,62 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal sealed class OperationMapBuilder : OperationWalker<Dictionary<SyntaxNode, IOperation>>
+    {
+        internal static readonly OperationMapBuilder Instance = new OperationMapBuilder();
+
+        private OperationMapBuilder() { }
+
+        public override object? Visit(IOperation? operation, Dictionary<SyntaxNode, IOperation> argument)
+        {
+            if (operation == null)
+            {
+                return null;
+            }
+
+            RecordOperation(operation, argument);
+
+            return base.Visit(operation, argument);
+        }
+
+        public override object? VisitBinaryOperator([DisallowNull] IBinaryOperation? operation, Dictionary<SyntaxNode, IOperation> argument)
+        {
+            // In order to handle very large nested operators, we implement manual iteration here. Our operations are not order sensitive,
+            // so we don't need to maintain a stack, just iterate through every level. Visit above will have already recorded the given operation.
+            Debug.Assert(argument.ContainsKey(operation.Syntax) || operation.IsImplicit);
```

Comment:
>Debug.Assert(argument.ContainsKey(operation.Syntax) || operation.IsImplicit); [](start = 12, length = 77)

This doesn't look like a safe assumption because the method can be called directly.

Reply:
With the private walker change, this can no longer be called directly.

---
In reply to: [526926610](https://github.com/dotnet/roslyn/pull/49453#discussion_r526926610) [](ancestors = 526926610)